### PR TITLE
Multi task

### PR DESCRIPTION
**Pull request by fangkuann**
support Multi-Task Learning using estimator.multi_head

### CODE REVIEW

File: tensorflow_ranking/python/model.py [hunk 1]
```diff
@@ -396,6 +402,94 @@ def _compute_logits_impl(self, context_features, example_features, labels,
     return logits
 
 
+class _MultiTaskGroupwiseRankingModel(_GroupwiseRankingModel):
```
Comment (xuanhuiwang) [inline tensorflow_ranking/python/model.py]: This is a very nice proposal. Indeed, we had a proposal internally and have been working on some implementation. Your PR is a very good reference for us to build a polished version.

Since this touches the core code, we would like this to be as concise as possible. We will also have multiple small changes to achieve the goal of multi-task learning.

Stayed tuned.

**Comment by xuanhuiwang** [comment]
@fangkuann, I've updated the github with the multi-task support in tf-ranking. Please see the latest few commits here: https://github.com/tensorflow/ranking/commits/master.

I also included a simple demo on how to use the multi-head in this commit: https://github.com/tensorflow/ranking/commit/1d91c5fd50cd8e855eaed6d8b69b0ec3f535c412. You may be able to adapt to your use case.

The main design is to have logits and labels as dicts that map head names to Tensors.

Please let us know if you find anything wrong.

**Comment by xuanhuiwang** [comment]
Closing this PR and let us know if you have other questions.

**Comment by fangkuann** [comment]
@xuanhuiwang 
Thanks for this great support! 
I will study the code and try this feature in my project.

**Comment by xuanhuiwang** [comment]
@fangkuann, let us know how your project goes and we are happy to provide more help.

**Comment by fangkuann** [comment]
> @fangkuann, let us know how your project goes and we are happy to provide more help.

I work as a search engineer mainly focused on optimizing search result for our product, which is the largest question-answer community like Quora in China. 
Recently, we are endeavoring to switch our ranking model from traditional GBDT to NN based model.
The TF-Ranking library greatly simplified our work. We use this library to train a DNN model and after a period's parameter tuning, the NN baseline beat the GBDT and was running on the fly very well till now.

Currently, we are trying several directions to further improve model performance, 

- multi-task learning
The baseline model was trained on user engagement like click. While we want to improve user satisfactory metrics(eg, upvote, collect, reading duration...) while maintaining user engagement performance. 
We have experimented shared bottom multi-task structure but did not achieve substantial improvement. We are going to try the MMOE model in the next stage.

- unbiased learning to rank
The TF-Ranking support weight_feature_name parameter to assign each instance sample weight. We use Position-Biased Click Model to obtain weight for each position.  And trained an unbiased model using sample weight manner. 
However, this manner degrades model performance seriously in offline evaluation, this is mainly because user clicks in the low position may contain much more noise, and enhance these sample's weight could let model miss the correct part of the training set. 
We are going to try another unbiased training method which factorizes user click into the biased and unbiased part, like this paper  https://dl.acm.org/citation.cfm?id=3298689.3347033. 
For unbiased learning, I am confusing how to properly evaluate model performance in offline. Since user cilck data collected is a biased dataset. Does evaluate unbiased ranking using biased dataset is appropriate? 

- model ensemble
For tabular dataset(mainly dense features, very less embedding features), the GBDT works quite well and NN may not beat GBDT. There are many papers tries to ensemble GBDT and NN. 
We have tried using Adanet, like this issue, but met some problem for integrating adanet and TF-Ranking https://github.com/tensorflow/adanet/issues/121.  Therefore this direction is suspended.

**Comment by xuanhuiwang** [comment]
@fangkuann, thanks a lot for your feedback. We are very glad to know that tf-ranking is used in your production!

Re: multi-task. Thanks for sharing the results. MMOE sounds good for the next step.
Re: unbiased learning-to-rank. Did you compare the models based on weighted metrics such as weighted MRR or weighted DCG? These are the better metrics than unweighted ones.
Re: model ensemble. We haven't tried anything with AdaNet yet. So far, we haven't planned on this yet.

**Comment by fangkuann** [comment]
Thanks for your advice! 
With weighted DCG metrics, we have much better offline performance now.  We will continue trying this excellent library for future development.
